### PR TITLE
Add IDNA punycode support in Cloudflare DNS API

### PR DESCRIPTION
**Pull request by mtju**
I have discovered that Cloudflare DNS method was not working with my punycode domains so I have made some changes and tested them with a variety of domains.

### CODE REVIEW

**Comment by mtju** [comment]
@Neilpang could you please consider merging this change?

File: dnsapi/dns_cf.sh [hunk 1]
```diff
@@ -197,3 +203,33 @@ _cf_rest() {
...
   return 0
 }
+
+# Check if Cloudflare's response name value contains given domain
+_cf_response_name_contains() {
+
+  if _is_idna_punycode "$1"; then
+    # Because name value in Cloudflare's IDNA punycode queries is
+    # decoded we also need to decode domain using idn
+    _decoded_domain=$(idn --idna-to-unicode "$1")
```
Comment (Neilpang) [inline dnsapi/dns_cf.sh]: we have functions `_is_idn()` and `_idn`,
please use that instead.

File: dnsapi/dns_cf.sh [hunk 2]
```diff
@@ -197,3 +203,33 @@ _cf_rest() {
...
+    fi
+  fi
+
+  return 1
+}
+
+# Check if supplied domain is in IDNA punycode format
+_is_idna_punycode() {
+
+  if echo "$1" | grep -E "(^|\.)xn--" >/dev/null 2>&1; then
```
Comment (Neilpang) [inline dnsapi/dns_cf.sh]: do not use `grep -E`,  use `_egrep_o()` function instead.

**Comment by Neilpang** [comment]
ponycode issue with cloudflare was already fixed.

**Pull request closed without merging**